nomos-da/network/core/src/dispersal/executor/behaviour.rs

```diff
@@ -0,0 +1,439 @@
+                error,
+                blob_id: blob_id.clone().try_into().unwrap(),
+                subnetwork_id,
+            })?;
+        stream
+            .stream
+            .flush()
+            .await
+            .map_err(|error| DispersalError::Io {
+                error,
+                blob_id: blob_id.clone().try_into().unwrap(),
+                subnetwork_id,
+            })?;
+        let response: DispersalRes =
+            unpack_from_reader(&mut stream.stream)
+                .await
+                .map_err(|error| DispersalError::Io {
+                    error,
+                    blob_id: blob_id.clone().try_into().unwrap(),
+                    subnetwork_id,
+                })?;
+        // Safety: blob_id should always be a 32bytes hash, currently is abstracted into a `Vec<u8>`
+        // but probably we should have a `[u8; 32]` wrapped in a custom type `BlobId`
+        // TODO: use blob_id when changing types to [u8; 32]
+        Ok((blob_id.try_into().unwrap(), subnetwork_id, response, stream))
+    }
+
+    /// Run when a stream gets free, if there is a pending task for the stream it will get scheduled to run
+    /// otherwise it is parked as idle.
+    fn handle_stream(
+        tasks: &mut FuturesUnordered<StreamHandlerFuture>,
+        to_disperse: &mut HashMap<PeerId, VecDeque<(SubnetworkId, DaBlob)>>,
+        idle_streams: &mut HashMap<PeerId, DispersalStream>,
+        stream: DispersalStream,
+    ) {
+        if let Some((subnetwork_id, next_request)) =
+            Self::next_request(&stream.peer_id, to_disperse)
+        {
+            let fut = Self::stream_disperse(stream, next_request, subnetwork_id).boxed();
+            tasks.push(fut);
+        } else {
+            // There is no pending request, so just idle the stream
+            idle_streams.insert(stream.peer_id, stream);
+        };
+    }
+
+    /// Get a pending request if its available
+    fn next_request(
+        peer_id: &PeerId,
+        to_disperse: &mut HashMap<PeerId, VecDeque<(SubnetworkId, DaBlob)>>,
+    ) -> Option<(SubnetworkId, DaBlob)> {
+        to_disperse
+            .get_mut(peer_id)
+            .and_then(|queue| queue.pop_front())
+    }
+}
+
+impl<Membership: MembershipHandler<Id = PeerId, NetworkId = SubnetworkId> + 'static>
+    DispersalExecutorBehaviour<Membership>
+{
+    /// Schedule a new task for sending the blob, if stream is not available queue messages for later
+    /// processing.
+    fn disperse_blob(
+        tasks: &mut FuturesUnordered<StreamHandlerFuture>,
+        idle_streams: &mut HashMap<Membership::Id, DispersalStream>,
+        membership: &mut Membership,
+        connected_subnetworks: &mut HashMap<PeerId, ConnectionId>,
+        to_disperse: &mut HashMap<PeerId, VecDeque<(Membership::NetworkId, DaBlob)>>,
+        subnetwork_id: SubnetworkId,
+        blob: DaBlob,
+    ) {
+        let members = membership.members_of(&subnetwork_id);
+        let peers = members
```

Comment:
I think to understand we are filtering here for already connected peers. In the hypothetical scenario (which seems more and more unlikely) that we have more members in a subnet than already connected peers - we would not disperse to them?

Reply:
The take here is that we are connected to at least one node in the subnetwork. If we are connected to more than one we try to send it as well yes.